[stable/redis] support redis docker library image

jlegrone opened the pull request:
<!--
Thank you for contributing to helm/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/helm/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/helm/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->

**What this PR does / why we need it**:
~This release switches the default image from [`docker.io/bitnami/redis`](https://hub.docker.com/r/bitnami/redis/) to [`docker.io/redis`](https://hub.docker.com/r/library/redis/). The bitnami image remains compatible with this chart; see `values-bitnami.yaml`.~

The default image in this release may be switched out for any image containing the `redis-server` binary. If `redis-server` is not the default image ENTRYPOINT, `master.command` must be specified. 

This allows `stable/redis` to offer a higher degree of flexibility for those who may need to run images containing [redis modules](https://redis.io/modules) or based on a different linux distribution than what is currently offered by bitnami.

#### Breaking changes:
- `master.args` and `slave.args` are removed. Use `master.command` or `slave.command` instead in order to override the image entrypoint, or `master.extraFlags` to pass additional flags to `redis-server`.
- `disableCommands` is now interpreted as an array of strings instead of a string of comma separated values.
- `master.persistence.path` now defaults to `/data`.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*:

Related: https://github.com/bitnami/bitnami-docker-redis/issues/92

**Special notes for your reviewer**:

Several interesting test cases:
- `helm upgrade --install redis-test ./stable/redis -f ./stable/redis/values-bitnami.yaml`
- `helm upgrade --install redis-test ./stable/redis --set usePassword=false`
- `helm upgrade --install redis-test ./stable/redis --set master.persistence.enabled=false`
- `helm upgrade --install redis-test ./stable/redis --set configmap="maxmemory-policy volatile-lru"`
- `helm upgrade --install redis-test ./stable/redis --set image.repository=redislabs/rejson --set image.tag=latest`

prydonius commented on stable/redis/values-production.yaml:
```diff
@@ -193,11 +189,12 @@ master:
   ## https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/#updating-statefulsets
   statefulset:
-    updateStrategy: OnDelete
+    updateStrategy: RollingUpdate
```
Why is this change needed? It is not described in the PR description.

jlegrone replied:
I copied and pasted my updates from `values.yaml`, and I think there were a few additional things that had gotten out of sync. This didn't look like a value that should be different in prod environments.

prydonius reviewed (commented):
This PR is doing two different things, that I think should be split into different PRs, and probably warrant some discussion.

1. Changing the default image to Docker Inc.'s Redis

I'm curious what the motivation for this change is -- is there something that Docker Inc.'s Redis image has that the Bitnami Redis image does not? In particular, changing the image of this chart has deeper consequences as it has automation setup to keep the bitnami/redis image up-to-date and secure (e.g. https://github.com/helm/charts/pull/7065).

2. Generalising the chart to support custom Redis images

I'm not fully convinced it makes sense for this chart to try to support every type of Redis image, they're all going to be a little bit different (env vars, volume paths, etc.). This sort of approach will make charts much more complex and harder to maintain and test. Instead, we can create different charts that are tailored to specific usecases (e.g. a redis chart that focuses on making it easy to add extra modules, or includes some more common modules vs a bare bones Redis chart). With the charts project moving to a more [distributed chart repositories model](https://github.com/helm/community/blob/master/proposals/distributed-search.md), it's going to be more common to see different variants of the same application for different usecases, and they can be hosted and maintained by anyone.

tompizmor commented on stable/redis/README.md:
```diff
@@ -80,12 +91,12 @@ The following table lists the configurable parameters of the Redis chart and the
 | `global.imageRegistry`                     | Global Docker image registry                                                                                   | `nil`                                                |
 | `image.registry`                           | Redis Image registry                                                                                           | `docker.io`                                          |
-| `image.repository`                         | Redis Image name                                                                                               | `bitnami/redis`                                      |
+| `image.repository`                         | Redis Image name                                                                                               | `redis`                                              |
```
I guess this is `bitnami/redis` again no?

jlegrone replied:
good catch, thanks

jlegrone committed:
refactor(stable/redis): support redis docker library image

BREAKING CHANGES:
- `master.args` and `slave.args` are removed. Use `master.command` or
  `slave.command` instead in order to override the image entrypoint,
   or `master.extraFlags` to pass additional flags to `redis-server`.
- `disableCommands` is now interpreted as an array of strings instead
   of a string of comma separated values.
- `master.persistence.path` now defaults to `/data`.

Signed-off-by: Jacob LeGrone <git@jacob.work>

jlegrone committed:
test(stable/redis): add ci values for base images & configurations

Signed-off-by: Jacob LeGrone <git@jacob.work>
